File: Data Structures & Algorithms/max-water-container/submission-0.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int n = heights.size();
        int left = 0;
        int right = n - 1;
        if (heights.size() == 0 || heights.size() == 1) return 0;
        int maxWater = 0;
        while (left < right){
           int water = (right - left) * min(heights[left], heights[right]); 
           maxWater = max(water, maxWater);
           if (heights[left] < heights[right]) left++;
           else right--; 
        }
        return maxWater;
    }
};

File: Data Structures & Algorithms/count-good-nodes-in-binary-tree/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int goodNodes(TreeNode root) {
        return dfs(root, Integer.MIN_VALUE);
    }
    private int dfs (TreeNode node, int largestVal) {
        if (node == null) return 0;
        int count = node.val >= largestVal ? 1 : 0;
        largestVal = Math.max(largestVal, node.val);
        return count + dfs(node.left, largestVal) + dfs(node.right, largestVal);
    }
}
